### DISCUSSION TITLE
The `GIT_INSTALL_ROOT` environment variable and detecting a git for windows installation

### DISCUSSION THREAD

**Discussion by ykhan21** [git-for-windows/git#4985] [Q&A]
`scoop install git` (git for windows) sets the GIT_INSTALL_ROOT environment variable to `%USERPROFILE%\scoop\apps\git\current`.
1. Is this variable used in any way by git bash or git?
2. What causes GIT_INSTALL_ROOT to get set? It looks like a normal git install to `C:\Program Files` with the installer does not set it. scoop sets it. Does it get set in any case by the installer or other package managers?
3. What are common paths one should check on a system for a git for windows installation? `%GIT_INSTALL_ROOT%`, `C:\Program Files\Git`, `%USERPROFILE%\scoop\apps\git\current\`, ...

**Comment by rimrul** [comment]
> `scoop install git` (git for windows) sets the GIT_INSTALL_ROOT environment variable to `%USERPROFILE%\scoop\apps\git\current`.
> 
> 1. Is this variable used in any way by git bash or git?

I don't think so. It [seems to be used by `cmder`](https://github.com/ScoopInstaller/Main/commit/97045616d8685c58596e4eb65b6be6110795c11f)

> 2. What causes GIT_INSTALL_ROOT to get set? 

https://github.com/ScoopInstaller/Main/blob/master/bucket/git.json#L60

> It looks like a normal git install to `C:\Program Files` with the installer does not set it. scoop sets it. Does it get set in any case by the installer or other package managers?

I think it only gets set by `scoop` and `cmder`

> 3. What are common paths one should check on a system for a git for windows installation? `%GIT_INSTALL_ROOT%`, `C:\Program Files\Git`, `%USERPROFILE%\scoop\apps\git\current\`, ...

`C:\Program Files\Git` and `C:\Program Files (x86)\Git`.

**Comment by ykhan21** [comment]
@rimrul
1. Can you use the git installer to install somewhere besides `C:\Program Files\Git` and `C:\Program Files (x86)\Git`?
2. Because there are at least three common paths git is installed to (`C:\Program Files\Git`, `C:\Program Files (x86)\Git`, and `%USERPROFILE%\scoop\apps\git\current`), would the git-for-windows project consider setting this environment variable to allow other programs to find git? The necessary git programs are added to the Windows PATH, but some other projects try to find the git installation to use its gnu utils.

**Reply by dscho** [reply]
> 1. Can you use the git installer to install somewhere besides `C:\Program Files\Git` and `C:\Program Files (x86)\Git`?

It is possible, but not recommended. See the output of running the installer with `/?`

> 2\. would the git-for-windows project consider setting this environment variable to allow other programs to find git?

The canonical way to retrieve the location into which Git for Windows is installed is:

```sh
powershell -command '(Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\Git_is1" -Name InstallLocation).InstallLocation'
```

The biggest problem with your ask is that it is not possible, in general, to provide a unique answer, because it is totally possible for multiple versions of Git for Windows to be installed in parallel, via Portable Git or MinGit.

> [...] some other projects try to find the git installation to use its gnu utils.

While it currently works, please do note that there is no guarantee that it will keep working. The purpose of [Git for Windows](https://git-for-windows.github.io/) is really to bring *Git* to Windows, after all. This means that we have to ship quite a few Unix-like tools, but it does not mean that we will include such tools unless they are needed for Git's own operations, or at least for historical reasons (needs of active contributors are sometimes a factor in deciding whether to include a tool or not). Each and any GNU utility that is no longer required by Git may be dropped without any previous announcement.

Having said that, if that risk is acceptable to you, you could use Git aliases creatively to call those GNU utilities. For example, to run a Bash command-line without modifying the `PATH` (and using the default installer option where only `git.exe` and friends are added to the `PATH`), you could call something like this:

```sh
git -c alias.sh='!sh' sh -c 'for i in $(seq 10 -1 1); do printf "count-down: %d\n" $i; done'
```

**Reply by ykhan21** [reply]
> The canonical way to retrieve the location into which Git for Windows is installed ...

Thanks. Just wondering, will this work if it was installed with scoop and cmder? And will this work for any user, regardless of what user installed git? scoop installs git local to the user ...
 
> The biggest problem with your ask is that it is not possible, in general, to provide a unique answer, because it is totally possible for multiple versions of Git for Windows to be installed in parallel, via Portable Git or MinGit.

What would be the problem if the main installer set the `GIT_INSTALL_ROOT` variable?

> run a Bash command-line without modifying the PATH

I see, but what I mean is that these "other programs" use the gnu utils by invoking the bash.exe that comes with Git for Windows shell and passing a bash script for the bash.exe to run.

**Reply by rimrul** [reply]
> > The canonical way to retrieve the location into which Git for Windows is installed ...
> 
> Thanks. Just wondering, will this work if it was installed with scoop and cmder? And will this work for any user, regardless of what user installed git? scoop installs git local to the user ...

No, this only works for Git for Windows installed via the installer, but scoop uses Portable Git.

**Reply by dscho** [reply]
> > run a Bash command-line without modifying the PATH
> 
> I see, but what I mean is that these "other programs" use the gnu utils by invoking the bash.exe that comes with Git for Windows shell and passing a bash script for the bash.exe to run.

That's even easier:

```powershell
git -c alias.sh='!sh' sh path/to/script arguments
```

(But I have to caution that Git for Windows is not meant as a replacement for MSYS2, Cygwin or WSL. Too many users would then ask way too much of way too few maintainer(s), Git for Windows is primarily catering toward _Git's_ needs, not intending to provide a full Unix experience on Windows, even if some users seem to mistake it for trying to be one.)

**Reply by ykhan21** [reply]
I see, thanks, @dscho, @rimrul. I will check if MSYS can be used for these projects. Two questions, though:
1) Does `git -c alias.sh='!sh' ...` set a temporary alias called sh?
2) If it is not too much to ask, regarding MSYS2, will `pacman -S git` get me roughly the same environment as Git for Windows' bash does? I think I will need to add MSYS2's git to the Windows' path and (part of) the Windows path to MSYS2's path, but are there any major differences?

**Reply by dscho** [reply]
> 1. Does `git -c alias.sh='!sh' ...` set a temporary alias called sh?

Yes, that's what it does.

> 2\. will `pacman -S git` get me roughly the same environment as Git for Windows' bash does?

Not quite. It will install the MSYS version of Git which is a tad slower than Git for Windows and follows more the Unix semantics than the Windows semantics, too.

**Reply by ykhan21** [reply]
> follows more the Unix semantics than the Windows semantics, too.

Is it at least like use Git for Windows' git through Git Bash? Or MSYS's git also different from that?

Also, what would be the drawback of having the main Git for Windows installer set %GIT_INSTALL_ROOT%? It would make referring to "git" easier, even when not using the bash it provides.
For example, right now, to set a Git Bash profile in Windows Terminal (or any other terminal such as IntelliJ's terminal or Wezterm or Alacritty), you need to set your shell to `C:/Program Files/Git/bin/bash.exe -l -i` and the icon to `C:/Program Files/Git/mingw64/share/git/git-for-windows.ico`. You need to change these paths if you installed GfW to a different folder such as the x86 folder. With %GIT_INSTALL_ROOT%, the shell and icon would just be `%GIT_INSTALL_ROOT%/bin/bash.exe -l -i` and `%GIT_INSTALL_ROOT%/mingw64/share/git/git-for-windows.ico`, respectively, and it even cover installing git through scoop, which is somewhat common.

**Reply by dscho** [reply]
> what would be the drawback of having the main Git for Windows installer set %GIT_INSTALL_ROOT%?

It would raise expectations to levels I do not want to support (and cannot, either: After a recent reorg at GitHub, I no longer have an entire team to help out with maintaining Git for Windows).

> With %GIT_INSTALL_ROOT%, the shell and icon would just be `%GIT_INSTALL_ROOT%/bin/bash.exe -l -i` and `%GIT_INSTALL_ROOT%/mingw64/share/git/git-for-windows.ico`, respectively, and it even cover installing git through scoop, which is somewhat common.

That's precisely the level of expectations that make maintaining Git for Windows utterly hard. The hard-coded path `mingw64` is an internal implementation detail, yet users pretend that this is a given and that changing it would "break Git for Windows" because they rely on things that are not actually promised.

I fear that introducing support for setting `GIT_INSTALL_ROOT` would only exacerbate the situation, making maintaining Git even more involved and cumbersome than it already is.

So no, I am not eager to add this.
